auth/range_perm_cache.go

```diff
@@ -47,59 +47,51 @@ func isRangeEqual(a, b *rangePerm) bool {
 	return bytes.Equal(a.begin, b.begin) && bytes.Equal(a.end, b.end)
 }
 
-// removeSubsetRangePerms removes any rangePerms that are subsets of other rangePerms.
-// If there are equal ranges, removeSubsetRangePerms only keeps one of them.
-// It returns a sorted rangePerm slice.
-func removeSubsetRangePerms(perms []*rangePerm) (newp []*rangePerm) {
+// mergeRangePerms merges adjacent rangePerms.
+func mergeRangePerms(perms []*rangePerm) (rs []*rangePerm) {
+	if len(perms) < 2 {
+		return perms
+	}
 	sort.Sort(RangePermSliceByBegin(perms))
-	var prev *rangePerm
-	for i := range perms {
-		if i == 0 {
-			prev = perms[i]
-			newp = append(newp, perms[i])
+
+	// initialize with first rangePerm
+	rs = append(rs, perms[0])
+
+	// merge in-place from 2nd
+	for i := 1; i < len(perms); i++ {
+		// skip duplicate range
+		if isRangeEqual(rs[len(rs)-1], perms[i]) {
 			continue
 		}
-		if isRangeEqual(perms[i], prev) {
+
+		// merge ["a", "") with ["a", "c")
+		if bytes.Equal(rs[len(rs)-1].begin, perms[i].begin) && len(rs[len(rs)-1].end) == 0 && len(perms[i].end) > 0 {
+			rs[len(rs)-1].end = getMax(rs[len(rs)-1].end, perms[i].end)
 			continue
 		}
-		if isSubset(perms[i], prev) {
+
+		// do not merge ["a", "b") with ["b", "")
+		if bytes.Equal(rs[len(rs)-1].end, perms[i].begin) && len(perms[i].end) == 0 {
+			rs = append(rs, perms[i])
 			continue
 		}
-		if isSubset(prev, perms[i]) {
-			prev = perms[i]
-			newp[len(newp)-1] = perms[i]
+
+		// rs[len(rs)-1].end < perms[i].begin; found beginning of next range
+		if bytes.Compare(rs[len(rs)-1].end, perms[i].begin) < 0 {
+			rs = append(rs, perms[i])
 			continue
 		}
-		prev = perms[i]
-		newp = append(newp, perms[i])
+
+		rs[len(rs)-1].end = getMax(rs[len(rs)-1].end, perms[i].end)
```

Comment:
```go
		// rp.end < perms[i].begin; found beginning of next range
		if bytes.Compare(rp.end, perms[i].begin) < 0 {
			if len(perms[i].end) != 0 {
				rp.end = perms[i].end
			}
			continue
		}

		rp.end = getMax(rp.end, perms[i].end)
```

still fails?


```
=== RUN   TestGetMergedPerms
--- FAIL: TestGetMergedPerms (0.00s)
	range_perm_cache_test.go:155: #4: result="[\x00 \x0f] ", want="[\x00 \x01] [\x02 \x03] [\x04 \a] [\b \x0f] "
	range_perm_cache_test.go:155: #5: result="[00 19] ", want="[00 08] [10 12] [18 19] "
	range_perm_cache_test.go:155: #10: result="[a e] ", want="[a c] [d e] "
	range_perm_cache_test.go:155: #11: result="[a f] ", want="[a b] [c d] [e f] "
	range_perm_cache_test.go:155: #12: result="[a f] ", want="[a b] [c d] [e f] "
	range_perm_cache_test.go:155: #14: result="[1 z] ", want="[1 9] [a z] "
	range_perm_cache_test.go:155: #16: result="[1 z] ", want="[1 9] [a z] "
	range_perm_cache_test.go:155: #17: result="[1 f] ", want="[1 9] [a f] "
	range_perm_cache_test.go:155: #19: result="[a ] ", want="[a ] [b ] "
	range_perm_cache_test.go:155: #22: result="[a c] [c ] ", want="[a ] [b c] [c ] [d ] "
```

This is because of this corner case where we do not want to merge

```
		{
			[]*rangePerm{makePerm("a", ""), makePerm("b", "")},
			[]*rangePerm{makePerm("a", ""), makePerm("b", "")},
		},
```